grease/src/Grease/Shape.hs

```diff
@@ -314,3 +338,206 @@ deriving instance (ShowExt ext tag, ShowF tag) => Show (ArgShapes ext tag tys)
+  padding :: (Integral a) => a -> MemShape w NoTag
+  padding w = Uninitialized (toBytes w)
+  -- if we dont know where to place members we have to fail/just place some bytes
+  buildMember :: Word64 -> MDwarf.Member -> Maybe (Word64, Seq.Seq (MemShape w NoTag))
+  buildMember loc mem =
+    ( \memLoc ->
+        do
+          padd <- Just (if memLoc == loc then Seq.empty else Seq.singleton (padding $ memLoc - loc))
+          membershape <- shapeOfTyApp $ MDwarf.memberType mem
+          let memByteSize = sum $ memShapeSize ?ptrWidth <$> membershape
+          let nextLoc = memLoc + fromIntegral memByteSize
+          Just (nextLoc, padd Seq.>< membershape)
+    )
+      =<< MDwarf.memberLoc mem
+  shapeOfTyApp :: MDwarf.TypeRef -> Maybe (Seq.Seq (MemShape w NoTag))
+  shapeOfTyApp x = shapeSeq =<< extractType sprog x
+
+  shapeSeq :: MDwarf.TypeApp -> Maybe (Seq.Seq (MemShape w NoTag))
+  shapeSeq (MDwarf.UnsignedIntType w) = ishape w
+  shapeSeq (MDwarf.SignedIntType w) = ishape w
+  shapeSeq MDwarf.SignedCharType = ishape (1 :: Int)
+  shapeSeq MDwarf.UnsignedCharType = ishape (1 :: Int)
+  -- Need modification to DWARF to collect counts properly + evaluate ops for ub
+  shapeSeq (MDwarf.ArrayType _elTy _ub) = Nothing
+  -- need to compute padding between els, infront of els and at end
+  shapeSeq (MDwarf.StructType sdecl) =
+    let structSize = MDwarf.structByteSize sdecl
+     in do
+          (currLoc, seqs) <-
+            foldM
+              ( \(currloc, seqs) mem ->
+                  do
+                    (nextLoc, additionalShapes) <- buildMember currloc mem
+                    Just (nextLoc, seqs Seq.>< additionalShapes)
+              )
+              (0, Seq.empty)
+              (MDwarf.structMembers sdecl)
+          endPad <- Just (if currLoc >= structSize then Seq.empty else Seq.singleton (padding $ structSize - currLoc))
+          Just $ (seqs Seq.>< endPad)
+  shapeSeq _ = Nothing
+
+intPtrShape :: Symbolic.SymArchConstraints arch => MC.ArchReg arch tp -> Maybe (C.Some (PtrShape ext w NoTag))
+intPtrShape reg =
+  case MT.typeRepr reg of
+    MT.BVTypeRepr w -> Just $ C.Some $ ShapePtrBV NoTag w
+    _ -> Nothing
+
+pointerShapeOfDwarf :: (HasPtrWidth w, Symbolic.SymArchConstraints arch) => ArchContext arch -> MC.ArchReg arch tp -> Subprogram -> MDwarf.TypeApp -> Maybe (C.Some (PtrShape ext w NoTag))
+pointerShapeOfDwarf _ r _ (MDwarf.SignedIntType _) = intPtrShape r
+pointerShapeOfDwarf _ r _ (MDwarf.UnsignedIntType _) = intPtrShape r
+pointerShapeOfDwarf _ _ sprog (MDwarf.PointerType _ tyRef) =
+  let mshape = constructPtrTarget sprog =<< extractType sprog =<< tyRef
+      pshape = ShapePtr NoTag (Offset (toBytes (0 :: Int))) <$> mshape
+   in (C.Some <$> pshape)
+pointerShapeOfDwarf _ _ _ _ = Nothing
+
+-- Is there really nothing available that looks like this?
+takeJust :: (a -> Maybe b) -> [a] -> [b]
+takeJust _ [] = []
+takeJust f (h : tl) =
+  case f h of
+    Nothing -> []
+    Just e -> e : takeJust f tl
```

Comment:
This is [`Data.Maybe.mapMaybe`](https://hackage.haskell.org/package/base-4.21.0.0/docs/Data-Maybe.html#v:mapMaybe).